algorithms/integration/kapton_2019_correction.py

```diff
@@ -0,0 +1,538 @@
+        return absorption_correction
+
+    def abs_correction_flex(self, s1_flex):
+        """ Compute the absorption correction using beers law. Takes in a flex array of s1 vectors, determines path lengths for each
+            and then determines absorption correction for each s1 vector """
+        kapton_path_mm=flex.double()
+        kapton_faces=self.faces
+        from dials.algorithms.integration import get_kapton_path_cpp
+        # new style, much faster
+        kapton_path_mm=get_kapton_path_cpp(kapton_faces, s1_flex)
+        # old style, really slow
+        #for s1 in s1_flex:
+        #  kapton_path_mm.append(self.get_kapton_path_mm(s1))
+        # determine absorption correction
+        if kapton_path_mm is not None:
+            absorption_correction = 1/flex.exp(-self.abs_coeff * kapton_path_mm) # unitless, >=1
+        return absorption_correction
+
+    def zone_of_point_in_kapton_shadow(self,s1):
+        """Given an s1 vector, determine whether it is in the increasing/decreasing or no kapton shadow regime"""
+        pass
+
+    def distance_of_point_from_line(self, r0, r1, r2):
+        """ Evaluates distance between point and a line between two points
+            Note that implementation ignores z dimension"""
+        x0,y0,z0=r0
+        x1,y1,z1=r1
+        x2,y2,z2=r2
+        num=(y2-y1)*x0-(x2-x1)*y0+x2*y1-y2*x1
+        denom=math.sqrt((y2-y1)*(y2-y1)+(x2-x1)*(x2-x1))
+        return abs(num)/denom
+
+    def abs_bounding_lines_in_mm(self, detector):
+        """ Return bounding lines of kapton """
+        # first get bounding directions from detector:
+        detz=flex.mean(flex.double([panel.get_origin()[2] for panel in detector]))
+        edges=[]
+        for ii,panel in enumerate(detector):
+            f_size,s_size=panel.get_image_size()
+            for point in [(0,0), (0,s_size), (f_size,0), (f_size, s_size)]:
+                x,y = panel.get_pixel_lab_coord(point)[0:2]
+                edges.append((x,y,detz))
+        # Use the idea that the corners of the detector are end points of the diagonal and will be the
+        # top 2 max dimension among all end points
+        dlist=flex.double()
+        dlist_idx=[]
+        n_edges=len(edges)
+        for ii in range(n_edges-1):
+            for jj in range(ii+1,n_edges):
+                pt_1=col(edges[ii])
+                pt_2=col(edges[jj])
+                distance=(pt_1-pt_2).length()
+                dlist.append(distance)
+                dlist_idx.append((ii,jj))
+        sorted_idx=flex.sort_permutation(dlist, reverse=True)
+
+        edge_pts=[edges[dlist_idx[sorted_idx[0]][0]], edges[dlist_idx[sorted_idx[1]][0]], edges[dlist_idx[sorted_idx[0]][1]], edges[dlist_idx[sorted_idx[1]][1]]]
+
+        self.detector_edges=edge_pts
+       # Now get the maximum extent of the intersection of the rays with the detector
+        all_ints=[]
+        kapton_path_list = []
+
+        def not_present_in_list(mylist, datapoint, tolerance):
```

Comment:
Another embedded function that could be standalone (my main objection to these is that if it's embedded then anyone reading your code needs to parse the whole function to check that it doesn't capture anything - the operating assumption should be that deep in a package there should a reason for it).